zebrad/src/components/inbound.rs

```diff
@@ -110,96 +131,131 @@ impl Service<zn::Request> for Inbound {
+            let mut awaiting_state = Setup::FailedInit;
+            mem::swap(&mut self.network, &mut awaiting_state);
+            if let Setup::AwaitingNetwork {
+                mut network_setup,
+                verifier,
+            } = awaiting_state
+            {
+                match network_setup.try_recv() {
+                    Ok((outbound, address_book)) => {
+                        let downloads = Box::pin(Downloads::new(
+                            Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
+                            Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT),
+                            self.state.clone(),
+                        ));
+                        self.network = Setup::Initialized {
+                            address_book,
+                            downloads,
+                        };
+                    }
+                    Err(TryRecvError::Empty) => {
+                        // There's no setup data yet, so keep waiting for it
+                        self.network = Setup::AwaitingNetwork {
+                            network_setup,
+                            verifier,
+                        };
+                    }
+                    Err(error @ TryRecvError::Closed) => {
+                        // Mark the service as failed, because network setup failed
+                        error!(?error, "inbound network setup failed");
+                        self.network = Setup::FailedRecv {
+                            failed_setup: network_setup,
+                        };
+                        return Poll::Ready(Err(error.into()));
+                    }
                 }
-                Err(e @ TryRecvError::Closed) => {
-                    // In this case, report that the service failed, and put the
-                    // failed oneshot back so we'll fail again in case
-                    // poll_ready is called after failure.
-                    self.network_setup = Some(rx);
-                    return Poll::Ready(Err(e.into()));
-                }
-            };
+            }
         }
 
-        // Clean up completed download tasks
-        if let Some(downloads) = self.downloads.as_mut() {
-            while let Poll::Ready(Some(_)) = downloads.as_mut().poll_next(cx) {}
+        // Unfortunately, we can't combine these matches into an exhaustive match statement,
+        // because they use mutable references, or they depend on the state we've just modified.
+
+        // Make sure we left the network setup in a valid state
+        if matches!(self.network, Setup::FailedInit) {
+            unreachable!("incomplete Inbound initialization");
+        }
+
+        // If network setup failed, report service failure
+        if let Setup::FailedRecv { failed_setup } = &mut self.network {
+            // TryRecvError is not cloneable, so we have to generate a new error from the oneshot,
```

Comment:
might be worth doing the same thing we did with `SharedPeerError` here.

Reply:
The `SharedPeerError` and `fail_with` code has been the source of a lot of subtle bugs.

So I think this way is actually cleaner, because the untidy code is contained in a few lines in a single enum variant and a single match arm. And there are specific comments in those places explaining what's going on.

Reply:
> The `SharedPeerError` and `fail_with` code has been the source of a lot of subtle bugs.

what bugs was `SharedPeerError` the cause of? I just meant taking a non cloneable error and wrapping it in an `Arc` and a newtype to then implement error for the wrapped clonable version of the original error. I don't see how this could introduce bugs, and I know it potentially fixes bugs where subsequent attempts to create an error identical to the original end up actually creating an unrelated error.

Reply:
Here are all the bugs related to `SharedPeerError`:
https://github.com/ZcashFoundation/zebra/issues/1610
https://github.com/ZcashFoundation/zebra/pull/1600
https://github.com/ZcashFoundation/zebra/issues/1599
https://github.com/ZcashFoundation/zebra/issues/1510
https://github.com/ZcashFoundation/zebra/pull/1531
https://github.com/ZcashFoundation/zebra/pull/81

Reply:
As far as I can tell all of those issues are caused by `fail_with` and way the error handling for the network `Client` was originally designed, and none of them have anything to do with the cloneable wrapper around PeerErrors.